Add SingleKey and IfEmptyObject

ash-zzz opened the pull request:
Closes #813 

I also added IfEmptyObject since that was a missing guard type that I needed to use for this

sindresorhus commented:
I think it should be named `SingleKeyObject`. The current name is a bit ambiguous and could make you single it fetches a single key from an object, for example.

sindresorhus commented on source/single-key.d.ts:
```diff
@@ -0,0 +1,26 @@
+ */
+
+// export type SingleKey<ObjectType> =
+// 	IsUnion<keyof ObjectType> extends true ? never : {} extends ObjectType ? never : ObjectType;
```
Leftover

ash-zzz replied:
Oops I thought I removed them all

sindresorhus commented on source/single-key.d.ts:
````diff
@@ -0,0 +1,26 @@
+```
+
+@category Object
+ */
````
```suggestion
*/
```

sindresorhus commented on source/single-key.d.ts:
```diff
@@ -0,0 +1,36 @@
+	},
+	values: [1, 2]
+});
+// Compilation error
```
Document the actual compilation here.

sindresorhus commented on source/single-key.d.ts:
```diff
@@ -0,0 +1,36 @@
+});
+
+processOperation({
+	operation: {
+		name: 'add'
+	},
```
Keep the example simple by not using a nested property here.

ash-zzz commented:
> I think it should be named `SingleKeyObject`. The current name is a bit ambiguous and could make you single it fetches a single key from an object, for example.

Agreed, I just renamed it and fixed the documentation

sindresorhus commented:
Thanks :)

jakeboone02 commented:
Thanks for doing this, @ash-zzz!

Should this work for types where the keys are unrestricted? If not, I'm not sure it satisfies my original request from #813, which was to declare an object type that requires exactly one key without restricting the key itself to any specific string or union of strings.

I tried `SingleKeyObject<Record<string, any>>` and `SingleKeyObject<{ [k: string]: any }>` but they both resolved to `never`. Something like `SingleKeyObject<Record<'union' | 'of' | 'strings', any>>` works fine.
